src/source/Ice/SocketConnection.c

```diff
@@ -155,37 +160,75 @@ STATUS socketConnectionSendData(PSocketConnection pSocketConnection, PBYTE pBuf,
+    CHK(pSocketConnection != NULL, STATUS_NULL_ARG);
+    CHK((pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP || pDestIp != NULL), STATUS_INVALID_ARG);
     CHK_WARN(!ATOMIC_LOAD_BOOL(&pSocketConnection->connectionClosed), STATUS_SOCKET_CONNECTION_CLOSED_ALREADY, "Failed to send data. Socket closed already");
 
     MUTEX_LOCK(pSocketConnection->lock);
     locked = TRUE;
 
     if (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP && pSocketConnection->secureConnection) {
-        // underlying BIO has been bound to socket so SSL_write sends the encrypted data.
-        // because we are using non-blocking socket, SSL_write may return SSL_ERROR_WANT_WRITE in which case
-        // SSL_write should be called again with the same buffer.
-        while(iterate) {
+        if (SSL_is_init_finished(pSocketConnection->pSsl)) {
+            /* Should have a valid buffer */
+            CHK(pBuf != NULL && bufLen > 0, STATUS_INVALID_ARG);
             sslRet = SSL_write(pSocketConnection->pSsl, pBuf, bufLen);
-            if (sslRet > 0 || (sslErr = SSL_get_error(pSocketConnection->pSsl, sslRet)) != SSL_ERROR_WANT_WRITE) {
-                iterate = FALSE;
-            } else {
-                THREAD_SLEEP(SSL_WRITE_RETRY_DELAY);
+            if (sslRet < 0){
+                sslErr = SSL_get_error(pSocketConnection->pSsl, sslRet);
+                switch (sslErr) {
+                    case SSL_ERROR_WANT_READ:
+                        /* explicit fall-through */
+                    case SSL_ERROR_WANT_WRITE:
+                        break;
+                    default:
+                        DLOGW("SSL_write failed with %s", ERR_error_string(sslErr, NULL));
+                        DLOGD("Close socket %d", pSocketConnection->localSocket);
+                        ATOMIC_STORE_BOOL(&pSocketConnection->connectionClosed, TRUE);
+                        break;
+                }
+
+                CHK(FALSE, STATUS_SEND_DATA_FAILED);
             }
         }
-        if (sslRet < 0 && sslErr != SSL_ERROR_WANT_READ) {
-            DLOGW("SSL_write failed with %s", ERR_error_string(sslErr, NULL));
+
+        wBioDataLen = (SIZE_T) BIO_get_mem_data(SSL_get_wbio(pSocketConnection->pSsl), &wBioBuffer);
+        CHK_ERR(wBioDataLen >= 0, STATUS_SEND_DATA_FAILED, "BIO_get_mem_data failed");
+        if (wBioDataLen > 0 && send(pSocketConnection->localSocket, wBioBuffer, wBioDataLen, NO_SIGNAL) < 0) {
+            DLOGE("send data failed with errno %s", strerror(errno));
+
+            if (!socketShouldRetry(errno)) {
+                DLOGD("Close socket %d", pSocketConnection->localSocket);
+                ATOMIC_STORE_BOOL(&pSocketConnection->connectionClosed, TRUE);
+            }
+
+            /* need to also reset BIO on negative path */
+            retStatus = STATUS_SEND_DATA_FAILED;
         }
 
+        /* reset bio to clear its content since it's already sent */
+        BIO_reset(SSL_get_wbio(pSocketConnection->pSsl));
```

Comment:
It's cleaner to handle this in the cleanup

Reply:
I dont quite get how it's cleaner. It makes sense locality wise because we did BIO_get_mem_data, we should clear the content from the BIO by doing BIO_reset.

Reply:
I disagree. The clean-up clause is entirely designed for that purpose. The entire codebase adheres to this constructs. If something needs to be cleaned up it's cleaned up in the cleanup clause. If the function is logically complex to not allow it then we need to see if the function needs to be broken down.

The main issue with premature cleanup is the code maintenance. Within the next few iterations, inevitably, code will be inserted that will make the logic of early cleanup more complex.

This all said, if you feel that it's very targeted and local then go ahead.

Reply:
yh it's very targeted to TCP TLS only. If we put it in cleanup, we need to do vigorous check that it was TCP TLS, and it wasnt bailing for thing like STATUS_NULL_ARG or STATUS_INVALID_ARG.